File: objectModel/TypeScript/Persistence/ModelJson/LocalEntityDeclarationPersistence.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { ModelJson } from '..';
import {
    CdmCorpusContext,
    CdmDataPartitionDefinition,
    CdmDocumentDefinition,
    CdmFolderDefinition,
    CdmImport,
    CdmLocalEntityDeclarationDefinition,
    CdmManifestDefinition,
    cdmObjectType,
    cdmLogCode,
    CdmTraitDefinition,
    CdmTraitReference,
    copyOptions,
    resolveOptions,
    traitToPropertyMap,
    CdmConstants
} from '../../internal';
import { Logger } from '../../Utilities/Logging/Logger';
import * as timeUtils from '../../Utilities/timeUtils';
import * as extensionHelper from './ExtensionHelper';
import { Entity, LocalEntity, Partition } from './types';

export class LocalEntityDeclarationPersistence {
    private static TAG: string = LocalEntityDeclarationPersistence.name;

    public static async fromData(
        ctx: CdmCorpusContext,
        documentFolder: CdmFolderDefinition,
        dataObj: LocalEntity,
        extensionTraitDefList: CdmTraitDefinition[],
        manifest: CdmManifestDefinition
    ): Promise<CdmLocalEntityDeclarationDefinition> {
        const localEntityDec: CdmLocalEntityDeclarationDefinition =
            ctx.corpus.MakeObject(cdmObjectType.localEntityDeclarationDef, dataObj.name);
        localEntityDec.virtualLocation = documentFolder.folderPath + CdmConstants.modelJsonExtension;

        const localExtensionTraitDefList: CdmTraitDefinition[] = [];
        const entityDoc: CdmDocumentDefinition = await ModelJson.DocumentPersistence.fromData(
            ctx,
            dataObj,
            extensionTraitDefList,
            localExtensionTraitDefList
        );

        documentFolder.documents.push(entityDoc);

        // Entity schema path is the path to the doc containing the entity definition.
        localEntityDec.entityPath = ctx.corpus.storage.createRelativeCorpusPath(`${entityDoc.atCorpusPath}/${dataObj.name}`, manifest);

        localEntityDec.explanation = dataObj.description;

        if (dataObj['cdm:lastFileStatusCheckTime'] !== undefined) {
            localEntityDec.lastFileStatusCheckTime = new Date(dataObj['cdm:lastFileStatusCheckTime']);
        }

        if (dataObj['cdm:lastFileModifiedTime'] !== undefined) {
            localEntityDec.lastFileModifiedTime = new Date(dataObj['cdm:lastFileModifiedTime']);

        }

        if (dataObj['cdm:lastChildFileModifiedTime'] !== undefined) {
            localEntityDec.lastChildFileModifiedTime = new Date(dataObj['cdm:lastChildFileModifiedTime']);
        }

        if (dataObj.isHidden === true) {
            const isHiddenTrait: CdmTraitReference = ctx.corpus.MakeObject(cdmObjectType.traitRef, 'is.hidden', true);
            isHiddenTrait.isFromProperty = true;
            localEntityDec.exhibitsTraits.push(isHiddenTrait);
        }

        // Add traits for schema entity info.
        if (dataObj.schemas) {
            const t2pm: traitToPropertyMap = new traitToPropertyMap(localEntityDec);
            t2pm.updatePropertyValue('cdmSchemas', dataObj.schemas);
        }

        // Data partitions are part of the local entity, add them here.
        if (dataObj.partitions) {
            for (const element of dataObj.partitions) {
                const cdmPartition: CdmDataPartitionDefinition =
                    await ModelJson.DataPartitionPersistence.fromData(
                        ctx,
                        element,
                        extensionTraitDefList,
                        localExtensionTraitDefList,
                        documentFolder
                    );
                if (cdmPartition !== undefined) {
                    localEntityDec.dataPartitions.push(cdmPartition);
                } else {
                    Logger.error(ctx, this.TAG, this.fromData.name, null, cdmLogCode.ErrPersistModelJsonDocConversionError);
                    return undefined;
                }
            }
        }

        const importDocs: CdmImport[] = await extensionHelper.standardImportDetection(
            ctx,
            extensionTraitDefList,
            localExtensionTraitDefList
        );
        extensionHelper.addImportDocsToManifest(ctx, importDocs, entityDoc);

        return localEntityDec;
    }

    public static async toData(
        instance: CdmLocalEntityDeclarationDefinition,
        manifest: CdmManifestDefinition,
        resOpt: resolveOptions,
        options: copyOptions
    ): Promise<LocalEntity> {
        // Fetch the document from entity schema.
        const localEntity: LocalEntity =
            await ModelJson.DocumentPersistence.toData(instance.entityPath, manifest, resOpt, options, instance.ctx);

        if (localEntity !== undefined) {
            const t2pm: traitToPropertyMap = new traitToPropertyMap(instance);
            const isHiddenTrait: CdmTraitReference = t2pm.fetchTraitReference('is.hidden');

            if (!localEntity.description) {
                localEntity.description = instance.explanation;
            }
            localEntity['cdm:lastFileStatusCheckTime'] = timeUtils.getFormattedDateString(instance.lastFileStatusCheckTime);
            localEntity['cdm:lastFileModifiedTime'] = timeUtils.getFormattedDateString(instance.lastFileModifiedTime);
            localEntity['cdm:lastChildFileModifiedTime'] = timeUtils.getFormattedDateString(instance.lastChildFileModifiedTime);

            if (isHiddenTrait !== undefined) {
                localEntity.isHidden = true;
            }

            const schemas: string[] = t2pm.fetchPropertyValue('cdmSchemas') as string[];
            if (schemas !== undefined) {
                localEntity.schemas = schemas;
            }

            if (instance.dataPartitions !== undefined && instance.dataPartitions.length > 0) {
                localEntity.partitions = [];
                for (const element of instance.dataPartitions) {
                    const partition: Partition = await ModelJson.DataPartitionPersistence.toData(element, resOpt, options);
                    if (partition !== undefined) {
                        localEntity.partitions.push(partition);
                    }
                }
            }

            return localEntity;
        }

        return undefined;
    }

    private static copyExtensionProperties(entityFrom: Entity, entityDest: LocalEntity): void {
        for (const prop of Object.keys(entityFrom)
            .filter((x: string) => x.indexOf(':') !== -1)) {
            entityDest[prop] = entityFrom[prop];
        }
    }
}

File: objectModel/Java/objectmodel/src/main/java/com/microsoft/commondatamodel/objectmodel/persistence/modeljson/LocalEntityDeclarationPersistence.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.persistence.modeljson;

import com.microsoft.commondatamodel.objectmodel.cdm.CdmCorpusContext;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmDataPartitionDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmEntityDeclarationDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmFolderDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmLocalEntityDeclarationDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmImport;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmManifestDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmTraitDefinition;
import com.microsoft.commondatamodel.objectmodel.cdm.CdmTraitReference;
import com.microsoft.commondatamodel.objectmodel.enums.CdmLogCode;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.enums.CdmPropertyName;
import com.microsoft.commondatamodel.objectmodel.persistence.PersistenceLayer;
import com.microsoft.commondatamodel.objectmodel.persistence.modeljson.types.LocalEntity;
import com.microsoft.commondatamodel.objectmodel.persistence.modeljson.types.Partition;
import com.microsoft.commondatamodel.objectmodel.utilities.CopyOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.ResolveOptions;
import com.microsoft.commondatamodel.objectmodel.utilities.TraitToPropertyMap;
import com.microsoft.commondatamodel.objectmodel.utilities.logger.Logger;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class LocalEntityDeclarationPersistence {
  private static final String TAG = LocalEntityDeclarationPersistence.class.getSimpleName();

  public static CompletableFuture<CdmEntityDeclarationDefinition> fromData(
      final CdmCorpusContext ctx,
      final CdmFolderDefinition documentFolder,
      final LocalEntity obj,
      final List<CdmTraitDefinition> extensionTraitDefList,
      final CdmManifestDefinition manifest) {
    final CdmLocalEntityDeclarationDefinition localEntity = ctx.getCorpus()
        .makeObject(CdmObjectType.LocalEntityDeclarationDef, obj.getName());

    final List<CdmTraitDefinition> localExtensionTraitDefList = new ArrayList<>();
    return DocumentPersistence.fromData(ctx, obj, extensionTraitDefList, localExtensionTraitDefList)
        .thenCompose(entityDoc -> {

          documentFolder.getDocuments().add(entityDoc);

          // CdmEntityDefinition schema path is the path to the doc containing the entity definition.
          localEntity.setEntityPath(
              ctx.getCorpus()
                  .getStorage()
                  .createRelativeCorpusPath(
                      entityDoc.getAtCorpusPath() + "/" + obj.getName(),
                      documentFolder));
          localEntity.setVirtualLocation(documentFolder.getFolderPath() + PersistenceLayer.modelJsonExtension);
          localEntity.setExplanation(obj.getDescription());
          localEntity.setLastChildFileModifiedTime(obj.getLastChildFileModifiedTime());
          localEntity.setLastFileModifiedTime(obj.getLastFileModifiedTime());
          localEntity.setLastFileStatusCheckTime(obj.getLastFileStatusCheckTime());

          if (obj.isHidden() != null && obj.isHidden()) {
            final CdmTraitReference isHiddenTrait =
                ctx.getCorpus().makeRef(
                    CdmObjectType.TraitRef,
                    "is.hidden",
                    true);
            isHiddenTrait.setFromProperty(true);
            localEntity.getExhibitsTraits().add(isHiddenTrait);
          }

          // Add traits for schema entity info.
          if (obj.getSchemas() != null) {
            final TraitToPropertyMap t2pm = new TraitToPropertyMap(localEntity);
            t2pm.updatePropertyValue(CdmPropertyName.CDM_SCHEMAS, obj.getSchemas());
          }

          // Data partitions are part of the local entity, add them here.
          final List<Partition> partitions = obj.getPartitions();
          if (partitions != null) {
            for (final Partition element : partitions) {
              final CdmDataPartitionDefinition cdmPartition =
                  DataPartitionPersistence.fromData(
                      ctx,
                      element,
                      extensionTraitDefList,
                      localExtensionTraitDefList,
                      documentFolder)
                      .join();
              if (cdmPartition != null) {
                localEntity.getDataPartitions().add(cdmPartition);
              } else {
                Logger.error(ctx, TAG, "fromData", null, CdmLogCode.ErrPersistModelJsonDocConversionError);
                return CompletableFuture.completedFuture(null);
              }
            }
          }

          final List<CdmImport> importDocs =
              ExtensionHelper.standardImportDetection(ctx, extensionTraitDefList, localExtensionTraitDefList).join();
          ExtensionHelper.addImportDocsToManifest(ctx, importDocs, entityDoc);
          return CompletableFuture.completedFuture(localEntity);
        });
  }

  public static CompletableFuture<LocalEntity> toData(
      final CdmEntityDeclarationDefinition instance,
      final CdmManifestDefinition manifest,
      final ResolveOptions resOpt,
      final CopyOptions options) {

    return DocumentPersistence.toData(instance.getEntityPath(), manifest, instance.getCtx(), resOpt, options)
        .thenCompose(localEntity -> {
          if (localEntity != null) {
            final TraitToPropertyMap t2pm = new TraitToPropertyMap(instance);
            final CdmTraitReference isHiddenTrait = t2pm.fetchTraitReference("is.hidden");

            if (localEntity.getDescription() == null) {
                localEntity.setDescription(instance.getExplanation());
            }
            localEntity.setLastChildFileModifiedTime(instance.getLastChildFileModifiedTime());
            localEntity.setLastFileModifiedTime(instance.getLastFileModifiedTime());
            localEntity.setLastFileStatusCheckTime(instance.getLastFileStatusCheckTime());

            if (isHiddenTrait != null) {
              localEntity.setHidden(true);
            }

            final Object propertyValues = t2pm.fetchPropertyValue(CdmPropertyName.CDM_SCHEMAS);
            if (propertyValues instanceof ArrayList<?>) {
              final ArrayList<String> schemas = (ArrayList<String>) propertyValues;
              localEntity.setSchemas(schemas);
            }

            if (localEntity.getTraits() != null) {
              localEntity.setTraits(localEntity.getTraits());
            }

            if (instance.getDataPartitions() != null && instance.getDataPartitions().getCount() > 0) {
              localEntity.setPartitions(new ArrayList<>());
              for (final CdmDataPartitionDefinition element : instance.getDataPartitions()) {
                final Partition partition = DataPartitionPersistence.toData(element, resOpt, options).join();

                if (partition != null) {
                  localEntity.getPartitions().add(partition);
                }
              }
            }

            return CompletableFuture.completedFuture(localEntity);
          }

          return CompletableFuture.completedFuture(null);
        });
  }
}
